grok() panics on "a|b" pattern when both a and b contain named grok patterns

A community user recently [shared](https://brimdata.slack.com/archives/CTSMAK6G7/p1706196396370159) a Zed program using `grok()` that causes zq to panic.

At b9e99c2cdccf74a349aad0326397c05db53a4a00, `grok()` panics when called with a pattern argument containing `a|b` where both a and b contain named grok patterns (i.e., `%{PATTERN:field_name}`).

```console
$ zq 'grok("(%{INT:a}|%{INT:b})", "1")'
panic: runtime error: slice bounds out of range [:-1]
goroutine 1 [running]:
runtime/debug.Stack()
	/opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/debug/stack.go:24 +0x64
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/noah/brimdata/zed/runtime/sam/op/catcher.go:25 +0x3c
panic({0x103d68880?, 0x140001a80c0?})
	/opt/homebrew/Cellar/go/1.21.5/libexec/src/runtime/panic.go:914 +0x218
github.com/brimdata/zed/pkg/grok.(*Pattern).ParseValues(0x140000dd200, {0x104a01de8, 0x1})
	/Users/noah/brimdata/zed/pkg/grok/grok.go:192 +0x20c
github.com/brimdata/zed/runtime/sam/expr/function.(*Grok).Call(0x140005c1b40, {0x10284d5c4?, 0x10554f2d0?}, {0x140005c1d80?, 0x10554b258?, 0x1051ac5b8?})
	/Users/noah/brimdata/zed/runtime/sam/expr/function/grok.go:45 +0x330
github.com/brimdata/zed/runtime/sam/expr.(*Call).Eval(0x140005c7b30, {0x10554f2d0?, 0x140005c1e00}, {{0x103e2b290?, 0x104b31ac0?}, 0x0?, 0x1400049f8c8?})
	/Users/noah/brimdata/zed/runtime/sam/expr/eval.go:799 +0x130
github.com/brimdata/zed/runtime/sam/op/yield.(*Op).Pull(0x1400065aff0, 0x0?)
	/Users/noah/brimdata/zed/runtime/sam/op/yield/yield.go:31 +0x34c
github.com/brimdata/zed/runtime/sam/op.(*Single).Pull(0x1400061fb90, 0xf8)
	/Users/noah/brimdata/zed/runtime/sam/op/mux.go:134 +0x3c
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull(0x8?, 0x10?)
	/Users/noah/brimdata/zed/runtime/sam/op/catcher.go:28 +0x60
github.com/brimdata/zed/runtime/exec.(*Query).Pull(0x1400049fab8?, 0x10?)
	/Users/noah/brimdata/zed/runtime/exec/query.go:59 +0x48
github.com/brimdata/zed/zbuf.CopyPuller({0x10554f198, 0x1400061fa10}, {0x103e251b8, 0x1400065b020})
	/Users/noah/brimdata/zed/zbuf/batch.go:162 +0x54
github.com/brimdata/zed/cli/zq.(*Command).Run(0x14000022a80, {0x140001b8010, 0x1, 0x1})
	/Users/noah/brimdata/zed/cli/zq/command.go:162 +0x5b4
github.com/brimdata/zed/pkg/charm.path.run({0x1400019c338?, 0x1, 0x1}, {0x140001b8010?, 0x1, 0x0?})
	/Users/noah/brimdata/zed/pkg/charm/path.go:11 +0x68
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0x140000726e8?, {0x140001b8010, 0x1, 0x1})
	/Users/noah/brimdata/zed/pkg/charm/charm.go:63 +0x40
main.main()
	/Users/noah/brimdata/zed/cmd/zq/main.go:11 +0x5c
```

It looks like this is not unique to `a|b` patterns (or, at least, the patterns the user can see). I was working on some docs improvements to explain how the "custom pattern" approach should be used instead of the "named capture" syntax that are both shown in Logstash's [docs](https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html#_custom_patterns). However, when I tried to run their simple custom pattern example in Zed, I triggered what looks like the same panic shown above.

```
$ zq -version
Version: v1.12.0-45-gb9e99c2c

$ echo '"Jan  1 06:25:43 mailserver14 postfix/cleanup[21403]: BEF25A72965: message-id=<20130101142543.5828399CCAF@mailserver14.example.com>"' |
zq -z 'yield grok("%{SYSLOGBASE} %{POSTFIX_QUEUEID:queue_id}: %{GREEDYDATA:syslog_message}", this, "POSTFIX_QUEUEID [0-9A-F]{10,11}")' -

panic: runtime error: slice bounds out of range [:-1]
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x5e
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/zed/runtime/sam/op/catcher.go:25 +0x3d
panic({0x25466e0?, 0xc0005bc048?})
	/usr/local/opt/go/libexec/src/runtime/panic.go:914 +0x21f
github.com/brimdata/zed/pkg/grok.(*Pattern).ParseValues(0xc000bee690, {0xc0006e6000, 0x82})
	/Users/phil/work/zed/pkg/grok/grok.go:192 +0x238
github.com/brimdata/zed/runtime/sam/expr/function.(*Grok).Call(0xc0003c1380, {0xc00041f870?, 0x100dc5d?}, {0xc00061d560, 0x36d2960?, 0xc000800000?})
	/Users/phil/work/zed/runtime/sam/expr/function/grok.go:45 +0x486
github.com/brimdata/zed/runtime/sam/expr.(*Call).Eval(0xc00061f0e0, {0x4aa39338?, 0xc0003c1640}, {{0x28bd150?, 0x36d2960?}, 0xc000800000?, 0x1011045?})
	/Users/phil/work/zed/runtime/sam/expr/eval.go:799 +0x155
github.com/brimdata/zed/runtime/sam/op/yield.(*Op).Pull(0xc00077e7b0, 0x5?)
	/Users/phil/work/zed/runtime/sam/op/yield/yield.go:31 +0x432
github.com/brimdata/zed/runtime/sam/op.(*Single).Pull(0xc0006928a0, 0x20)
	/Users/phil/work/zed/runtime/sam/op/mux.go:134 +0x2b
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull(0x3cb6a68?, 0x50?)
	/Users/phil/work/zed/runtime/sam/op/catcher.go:28 +0x62
github.com/brimdata/zed/runtime/exec.(*Query).Pull(0x4aa39318?, 0xd8?)
	/Users/phil/work/zed/runtime/exec/query.go:59 +0x3e
github.com/brimdata/zed/zbuf.CopyPuller({0x4aa39318, 0xc0006926c0}, {0x28b70a0, 0xc00077e7e0})
	/Users/phil/work/zed/zbuf/batch.go:162 +0x4d
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc0004e2a80, {0xc0000400a0, 0x2, 0x2})
	/Users/phil/work/zed/cli/zq/command.go:162 +0x79e
github.com/brimdata/zed/pkg/charm.path.run({0xc0005a62a8?, 0x1, 0x1}, {0xc0000400a0?, 0x2, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x79
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0x100874b?, {0xc000040090, 0x3, 0x3})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x3f
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5b
```

--- a/pkg/grok/grok.go
+++ b/pkg/grok/grok.go
@@ -161,9 +161,10 @@ func (h Host) Compile(expr string) (*Pattern, error) {
 
 type Pattern struct {
 	*regexp.Regexp
-	s     map[string]int
-	order map[int]string
-	cache []string
+	s        map[string]int
+	order    map[int]string
+	keyCache []string
+	valCache []string
 }
 
 // Parse returns a map of matches on the input. The map can be empty.
@@ -179,19 +180,24 @@ func (p *Pattern) Parse(input string) map[string]string {
 	return r
 }
 
-func (p *Pattern) ParseValues(input string) []string {
+func (p *Pattern) ParseKeyValues(input string) ([]string, []string) {
 	a := p.FindStringSubmatchIndex(input)
 	if a == nil {
-		return nil
+		return nil, nil
 	}
-	p.cache = p.cache[:0]
-	for i := 0; len(p.cache) < len(p.s); i++ {
+	p.valCache = p.valCache[:0]
+	p.keyCache = p.keyCache[:0]
+	for i := 0; i < len(a)/2; i++ {
 		if _, ok := p.order[i]; !ok {
 			continue
 		}
-		p.cache = append(p.cache, input[a[i*2]:a[i*2+1]])
+		if a[i*2] == -1 {
+			continue
+		}
+		p.keyCache = append(p.keyCache, p.order[i])
+		p.valCache = append(p.valCache, input[a[i*2]:a[i*2+1]])
 	}
-	return p.cache
+	return p.keyCache, p.valCache
 }
 
 // Names returns all names that this pattern has in order.

--- a/pkg/grok/pattern_test.go
+++ b/pkg/grok/pattern_test.go
@@ -53,11 +53,11 @@ func TestPattern_Names(t *testing.T) {
 	require.Equal(t, []string{"zero", "three", "one", "two"}, p.Names())
 }
 
-func TestPattern_ParseValues(t *testing.T) {
+func TestPattern_ParseKeyValues(t *testing.T) {
 	h := NewBase()
 	p, err := h.Compile("%{TIMESTAMP_ISO8601:event_time} %{LOGLEVEL:log_level} %{GREEDYDATA:log_message}")
 	require.NoError(t, err)
-	ss := p.ParseValues("2020-09-16T04:20:42.45+01:00 DEBUG This is a sample debug log message")
+	_, ss := p.ParseKeyValues("2020-09-16T04:20:42.45+01:00 DEBUG This is a sample debug log message")
 	require.Equal(t, []string{"2020-09-16T04:20:42.45+01:00", "DEBUG", "This is a sample debug log message"}, ss)
 }
 
@@ -75,4 +75,13 @@ func TestPattern_NamesNested(t *testing.T) {
 	p, err := h.Compile("%{ONE:num.one}-%{ONE:[num][two]}")
 	require.NoError(t, err)
 	require.Equal(t, []string{"num.one", "[num][two]"}, p.Names())
+}
+
+func TestPattern_OptionalValues(t *testing.T) {
+	h := NewBase()
+	p, err := h.Compile("(%{INT:a}|%{INT:b})")
+	require.NoError(t, err)
+	keys, values := p.ParseKeyValues("1")
+	require.Equal(t, []string{"a"}, keys)
+	require.Equal(t, []string{"1"}, values)
 }

--- a/runtime/sam/expr/function/grok.go
+++ b/runtime/sam/expr/function/grok.go
@@ -12,6 +12,8 @@ type Grok struct {
 	zctx    *zed.Context
 	builder zcode.Builder
 	hosts   map[string]*host
+	// fields is used as a scratch space to avoid allocating a new slice.
+	fields []zed.Field
 }
 
 func newGrok(zctx *zed.Context) *Grok {
@@ -21,10 +23,10 @@ func newGrok(zctx *zed.Context) *Grok {
 	}
 }
 
-func (g *Grok) Call(_ zed.Allocator, vals []zed.Value) zed.Value {
-	patternArg, inputArg, defArg := vals[0], vals[1], zed.NullString
-	if len(vals) == 3 {
-		defArg = vals[2]
+func (g *Grok) Call(_ zed.Allocator, args []zed.Value) zed.Value {
+	patternArg, inputArg, defArg := args[0], args[1], zed.NullString
+	if len(args) == 3 {
+		defArg = args[2]
 	}
 	switch {
 	case zed.TypeUnder(defArg.Type()) != zed.TypeString:
@@ -38,19 +40,24 @@ func (g *Grok) Call(_ zed.Allocator, vals []zed.Value) zed.Value {
 	if err != nil {
 		return g.error(err.Error(), defArg)
 	}
-	p, err := h.getPattern(g.zctx, patternArg.AsString())
+	p, err := h.getPattern(patternArg.AsString())
 	if err != nil {
 		return g.error(err.Error(), patternArg)
 	}
-	ss := p.ParseValues(inputArg.AsString())
-	if ss == nil {
+	keys, vals := p.ParseKeyValues(inputArg.AsString())
+	if vals == nil {
 		return g.error("value does not match pattern", inputArg)
 	}
+	g.fields = g.fields[:0]
+	for _, key := range keys {
+		g.fields = append(g.fields, zed.NewField(key, zed.TypeString))
+	}
+	typ := g.zctx.MustLookupTypeRecord(g.fields)
 	g.builder.Reset()
-	for _, s := range ss {
+	for _, s := range vals {
 		g.builder.Append([]byte(s))
 	}
-	return zed.NewValue(p.typ, g.builder.Bytes())
+	return zed.NewValue(typ, g.builder.Bytes())
 }
 
 func (g *Grok) error(msg string, val zed.Value) zed.Value {
@@ -60,7 +67,7 @@ func (g *Grok) error(msg string, val zed.Value) zed.Value {
 func (g *Grok) getHost(defs string) (*host, error) {
 	h, ok := g.hosts[defs]
 	if !ok {
-		h = &host{Host: grok.NewBase(), patterns: make(map[string]*pattern)}
+		h = &host{Host: grok.NewBase(), patterns: make(map[string]*grok.Pattern)}
 		if err := h.AddFromReader(strings.NewReader(defs)); err != nil {
 			return nil, err
 		}
@@ -71,31 +78,18 @@ func (g *Grok) getHost(defs string) (*host, error) {
 
 type host struct {
 	grok.Host
-	patterns map[string]*pattern
+	patterns map[string]*grok.Pattern
 }
 
-func (h *host) getPattern(zctx *zed.Context, patternArg string) (*pattern, error) {
+func (h *host) getPattern(patternArg string) (*grok.Pattern, error) {
 	p, ok := h.patterns[patternArg]
 	if !ok {
-		pat, err := h.Host.Compile(patternArg)
+		var err error
+		p, err = h.Host.Compile(patternArg)
 		if err != nil {
 			return nil, err
 		}
-		var fields []zed.Field
-		for _, name := range pat.Names() {
-			fields = append(fields, zed.NewField(name, zed.TypeString))
-		}
-		typ, err := zctx.LookupTypeRecord(fields)
-		if err != nil {
-			return nil, err
-		}
-		p = &pattern{Pattern: pat, typ: typ}
 		h.patterns[patternArg] = p
 	}
 	return p, nil
-}
-
-type pattern struct {
-	*grok.Pattern
-	typ zed.Type
 }

